Enabling nullable context causes operator resolve errors

**Version Used**: 

```
Branch master (17 Apr 2019)
Latest commit 88d7d9 by Andrew Hall:
Merge pull request #34915 Add IVsRefactorNotify changes to MoveToNamespace refactorings

If available, use the host service to provide IVsRefactorNotify changes of the symbols being changed so dependent frameworks

can handle the changes appropriately.

Adds unit tests that verify we have the correct code operations produced when we expect symbols to change.
```

**Steps to Reproduce**:

```cs
#nullable enable
class C<T>
{
  public static extern C<T> operator &(C<T> c1, C<T> c2);
}

class Test
{
  // uncomment the following line to remove the error
  //#nullable disable
  void M(C<string> c1, string s)
  {
    var c2 = GetC(s);
    var c3 = c1 & c2; // CS0034
  }
    
  public static extern C<T> GetC<T>(T t);
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABAJgoGEAeAFQD5GA3owoUmAZlEBGAGxUAHvAS4OPXhQD2AB0QQY6hBQBkACi58KAY0lIV5iywCUAbkYBfRo1YVucWIOEUAPSBFPi4FupgYNQwFDAAFpQAZurE6gDuAJa4AOYUhNmUehQIcGDqAG5FiVQICPoBwXhEJGSUKJlQtHABTAAsFACyppxMkgAMalY2Y+MUUA4BQgwiIhUQBvYUALwUAOJwMOzGCy4rq+ubErtWRpYsTkEh7ADK4+NifQHu5yK9EmNZHAFIhlGY1AcjqpjNw4s43EA=

**Expected Behavior**:
No errors. The code compiles if you either remove the `#nullable enable` directive or uncomment the `#nullable disasble` directive

**Actual Behavior**:
```
error CS0034: Operator '&' is ambiguous on operands of type 'C<string>' and 'C<string>'
```
I'm not sure what exactly becomes ambiguous if nullable reference types are enabled. Both participating operands as well as their type arguments are explicitly non-nullable.

--- a/src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs
@@ -1,5 +1,6 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
@@ -746,10 +747,10 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                     // Return types must match exactly, parameters might match modulo identity conversion.
                     if (op.Signature.Kind == existingSignature.Kind && // Easy out
-                        op.Signature.ReturnType.Equals(existingSignature.ReturnType) &&
-                        op.Signature.LeftType.Equals(existingSignature.LeftType, TypeCompareKind.IgnoreDynamic) &&
-                        op.Signature.RightType.Equals(existingSignature.RightType, TypeCompareKind.IgnoreDynamic) &&
-                        op.Signature.Method.ContainingType.Equals(existingSignature.Method.ContainingType, TypeCompareKind.IgnoreDynamic))
+                        equalsIgnoringNullable(op.Signature.ReturnType, existingSignature.ReturnType) &&
+                        equalsIgnoringNullableAndDynamic(op.Signature.LeftType, existingSignature.LeftType) &&
+                        equalsIgnoringNullableAndDynamic(op.Signature.RightType, existingSignature.RightType) &&
+                        equalsIgnoringNullableAndDynamic(op.Signature.Method.ContainingType, existingSignature.Method.ContainingType))
                     {
                         equivalentToExisting = true;
                         break;
@@ -761,6 +762,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                     result.Add(op);
                 }
             }
+
+            static bool equalsIgnoringNullable(TypeSymbol a, TypeSymbol b) => a.Equals(b, TypeCompareKind.AllNullableIgnoreOptions);
+            static bool equalsIgnoringNullableAndDynamic(TypeSymbol a, TypeSymbol b) => a.Equals(b, TypeCompareKind.AllNullableIgnoreOptions | TypeCompareKind.IgnoreDynamic);
         }
 
         private bool GetUserDefinedOperators(

--- a/src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs
@@ -3,7 +3,6 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.ComponentModel;
 using System.Diagnostics;
 using System.Linq;
 using System.Threading;

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
@@ -97356,5 +97356,489 @@ class Program
                 // [A(default(string).F)]
                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOfTargetDelegate, "default(string).F").WithArguments("string? E.F<string?>(string? t)", "D<string>").WithLocation(10, 4));
         }
+
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_01()
+        {
+            var source =
+@"class C<T>
+{
+    public static C<T> operator &(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M(C<string> a, string s)
+    {
+        var b = Create(s);
+        _ = a & b;
+    }
+    static C<T> Create<T>(T t) => new C<T>();
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+        }
+
+        // C<T~> operator op(C<T~>, C<T~>)
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_02()
+        {
+            var source =
+@"#nullable disable
+class C<T>
+{
+    public static C<T> operator +(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+    {
+        _ = a + a;
+        _ = a + b;
+        _ = b + a;
+        _ = b + b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> c,
+#nullable enable
+        C<T> d)
+        where T : class?
+    {
+        _ = c + c;
+        _ = c + d;
+        _ = d + c;
+        _ = d + d;
+    }
+    static void M3<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x + x;
+        _ = x + y;
+        _ = x + z;
+        _ = y + x;
+        _ = y + y;
+        _ = y + z;
+        _ = z + x;
+        _ = z + y;
+        _ = z + z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (44,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator +(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y + z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator +(C<T> a, C<T> b)").WithLocation(44, 17),
+                // (46,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator +(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z + y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator +(C<T?> a, C<T?> b)").WithLocation(46, 17));
+        }
+
+        // C<T> operator op(C<T>, C<T>)
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_03()
+        {
+            var source =
+@"#nullable enable
+class C<T>
+{
+    public static C<T> operator -(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+    {
+        _ = a - a;
+        _ = a - b;
+        _ = b - a;
+        _ = b - b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> c,
+#nullable enable
+        C<T> d)
+        where T : class?
+    {
+        _ = c - c;
+        _ = c - d;
+        _ = d - c;
+        _ = d - d;
+    }
+    static void M3<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x - x;
+        _ = x - y;
+        _ = x - z;
+        _ = y - x;
+        _ = y - y;
+        _ = y - z;
+        _ = z - x;
+        _ = z - y;
+        _ = z - z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (44,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator -(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y - z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator -(C<T> a, C<T> b)").WithLocation(44, 17),
+                // (46,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator -(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z - y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator -(C<T?> a, C<T?> b)").WithLocation(46, 17));
+        }
+
+        // C<T~> operator op(C<T~>, C<T>)
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_04()
+        {
+            var source =
+@"class C<T>
+{
+#nullable disable
+    public static C<T> operator *(
+        C<T> a,
+#nullable enable
+        C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+    {
+        _ = a * a;
+        _ = a * b;
+        _ = b * a;
+        _ = b * b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> c,
+#nullable enable
+        C<T> d)
+        where T : class?
+    {
+        _ = c * c;
+        _ = c * d;
+        _ = d * c;
+        _ = d * d;
+    }
+    static void M3<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x * x;
+        _ = x * y;
+        _ = x * z;
+        _ = y * x;
+        _ = y * y;
+        _ = y * z;
+        _ = z * x;
+        _ = z * y;
+        _ = z * z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (47,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator *(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y * z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator *(C<T> a, C<T> b)").WithLocation(47, 17),
+                // (49,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator *(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z * y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator *(C<T?> a, C<T?> b)").WithLocation(49, 17));
+        }
+
+        // C<T~> operator op(C<T~>, C<T~>) where T : class?
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_05()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class?
+{
+#nullable disable
+    public static C<T> operator +(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+        where T : class?
+    {
+        _ = a + a;
+        _ = a + b;
+        _ = b + a;
+        _ = b + b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x + x;
+        _ = x + y;
+        _ = x + z;
+        _ = y + x;
+        _ = y + y;
+        _ = y + z;
+        _ = z + x;
+        _ = z + y;
+        _ = z + z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (34,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator +(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y + z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator +(C<T> a, C<T> b)").WithLocation(34, 17),
+                // (36,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator +(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z + y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator +(C<T?> a, C<T?> b)").WithLocation(36, 17));
+        }
+
+        // C<T!> operator op(C<T!>, C<T!>) where T : class?
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_06()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class?
+{
+    public static C<T> operator -(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+        where T : class?
+    {
+        _ = a - a;
+        _ = a - b;
+        _ = b - a;
+        _ = b - b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x - x;
+        _ = x - y;
+        _ = x - z;
+        _ = y - x;
+        _ = y - y;
+        _ = y - z;
+        _ = z - x;
+        _ = z - y;
+        _ = z - z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (33,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator -(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y - z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator -(C<T> a, C<T> b)").WithLocation(33, 17),
+                // (35,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator -(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z - y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator -(C<T?> a, C<T?> b)").WithLocation(35, 17));
+        }
+
+        // C<T~> operator op(C<T!>, C<T~>) where T : class?
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_07()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class?
+{
+#nullable disable
+    public static C<T> operator *(
+#nullable enable
+        C<T> a,
+#nullable disable
+        C<T> b) => a;
+}
+class Program
+{
+    static void M1<T>(
+#nullable disable
+        C<T> a,
+#nullable enable
+        C<T> b)
+        where T : class?
+    {
+        _ = a * a;
+        _ = a * b;
+        _ = b * a;
+        _ = b * b;
+    }
+    static void M2<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y,
+        C<T?> z)
+        where T : class
+    {
+        _ = x * x;
+        _ = x * y;
+        _ = x * z;
+        _ = y * x;
+        _ = y * y;
+        _ = y * z;
+        _ = z * x;
+        _ = z * y;
+        _ = z * z;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (38,17): warning CS8620: Argument of type 'C<T?>' cannot be used for parameter 'b' of type 'C<T>' in 'C<T> C<T>.operator *(C<T> a, C<T> b)' due to differences in the nullability of reference types.
+                //         _ = y * z;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "z").WithArguments("C<T?>", "C<T>", "b", "C<T> C<T>.operator *(C<T> a, C<T> b)").WithLocation(38, 17),
+                // (40,17): warning CS8620: Argument of type 'C<T>' cannot be used for parameter 'b' of type 'C<T?>' in 'C<T?> C<T?>.operator *(C<T?> a, C<T?> b)' due to differences in the nullability of reference types.
+                //         _ = z * y;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "y").WithArguments("C<T>", "C<T?>", "b", "C<T?> C<T?>.operator *(C<T?> a, C<T?> b)").WithLocation(40, 17));
+        }
+
+        // C<T~> operator op(C<T~>, C<T~>) where T : class
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_08()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class
+{
+#nullable disable
+    public static C<T> operator /(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y)
+        where T : class
+    {
+        _ = x / x;
+        _ = x / y;
+        _ = y / x;
+        _ = y / y;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        // C<T!> operator op(C<T!>, C<T!>) where T : class
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_09()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class
+{
+    public static C<T> operator &(C<T> a, C<T> b) => a;
+}
+class Program
+{
+    static void M<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y)
+        where T : class
+    {
+        _ = x & x;
+        _ = x & y;
+        _ = y & x;
+        _ = y & y;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        // C<T~> operator op(C<T!>, C<T~>) where T : class
+        [Fact]
+        [WorkItem(35057, "https://github.com/dotnet/roslyn/issues/35057")]
+        public void Operator_10()
+        {
+            var source =
+@"#nullable enable
+class C<T> where T : class
+{
+#nullable disable
+    public static C<T> operator |(
+#nullable enable
+        C<T> a,
+#nullable disable
+        C<T> b) => a;
+}
+class Program
+{
+    static void M<T>(
+#nullable disable
+        C<T> x,
+#nullable enable
+        C<T> y)
+        where T : class
+    {
+        _ = x | x;
+        _ = x | y;
+        _ = y | x;
+        _ = y | y;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
     }
 }

